File: examples/governance/compliance_domain/lib/compliance_domain/exemption/commands/approve_exemption.rb
module ComplianceDomain
  class Exemption
    module Commands
      class ApproveExemption
        include Hecks::Command
        emits "ApprovedExemption"

        attr_reader :exemption_id
        attr_reader :approved_by_id
        attr_reader :expires_at

        def initialize(
          exemption_id: nil,
          approved_by_id: nil,
          expires_at: nil
        )
          @exemption_id = exemption_id
          @approved_by_id = approved_by_id
          @expires_at = expires_at
        end

        def call
          existing = repository.find(exemption_id)
          if existing
            unless existing.status == "requested"
              raise Hecks::Error, "Cannot ApproveExemption: status must be 'requested', got '#{existing.status}'"
            end
            Exemption.new(
              id: existing.id,
              model_id: existing.model_id,
              policy_id: existing.policy_id,
              requirement: existing.requirement,
              reason: existing.reason,
              approved_by_id: approved_by_id,
              expires_at: expires_at,
              scope: existing.scope,
              approved_at: Time.now.to_s,
              status: "active"
            )
          else
            raise Hecks::Error, "Exemption not found: #{exemption_id}"
          end
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/approve_exemption.go
package domain

import (
	"time"
	"fmt"
)

type ApproveExemption struct {
	ExemptionId string `json:"exemption_id"`
	ApprovedById string `json:"approved_by_id"`
	ExpiresAt time.Time `json:"expires_at"`
}

func (c ApproveExemption) CommandName() string { return "ApproveExemption" }

func (c ApproveExemption) Execute(repo ExemptionRepository) (*Exemption, *ApprovedExemption, error) {
	existing, err := repo.Find(c.ExemptionId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("Exemption not found: %s", c.ExemptionId)
	}
	existing.ApprovedById = c.ApprovedById
	existing.ExpiresAt = c.ExpiresAt
	if existing.Status != "requested" {
		return nil, nil, fmt.Errorf("cannot ApproveExemption: Exemption is in %s state", existing.Status)
	}
	existing.Status = "active"
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := ApprovedExemption{
		AggregateID: existing.ID,
		ExemptionId: c.ExemptionId,
		ApprovedById: c.ApprovedById,
		ExpiresAt: c.ExpiresAt,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}
